src/credentials/FabricTable.cpp

```diff
@@ -1248,97 +1383,548 @@ CHIP_ERROR FabricTable::AllocatePendingOperationalKey(Optional<FabricIndex> fabr
+    return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR FabricTable::ActivatePendingOperationalKey(const Crypto::P256PublicKey & nocSubjectPublicKey)
 {
     // We can only manage commissionable pending fail-safe state if we have a keystore
     VerifyOrReturnError(mOperationalKeystore != nullptr, CHIP_ERROR_INCORRECT_STATE);
 
-    VerifyOrReturnError(!mIsPendingFabricDataPresent, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mStateFlags.Has(StateFlags::kIsOperationalKeyPending), CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnError(IsValidFabricIndex(mFabricIndexWithPendingState), CHIP_ERROR_INCORRECT_STATE);
 
-    CHIP_ERROR err = mOperationalKeystore->ActivateOpKeypairForFabric(mFabricIndexWithPendingState, nocSubjectPublicKey);
+    ReturnErrorOnFailure(mOperationalKeystore->ActivateOpKeypairForFabric(mFabricIndexWithPendingState, nocSubjectPublicKey));
+    mStateFlags.Clear(StateFlags::kIsOperationalKeyPending);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR FabricTable::AddNewPendingTrustedRootCert(const ByteSpan & rcac)
+{
+    VerifyOrReturnError(mOpCertStore != nullptr, CHIP_ERROR_INCORRECT_STATE);
 
-    if (err == CHIP_NO_ERROR)
+    // We should not already have pending NOC chain elements when we get here
+    ReturnErrorCodeIf(
+        mStateFlags.HasAny(StateFlags::kIsTrustedRootPending, StateFlags::kIsUpdatePending, StateFlags::kIsAddPending),
+        CHIP_ERROR_INCORRECT_STATE);
+
+    EnsureNextAvailableFabricIndexUpdated();
+    FabricIndex fabricIndexToUse = kUndefinedFabricIndex;
+
+    if (mNextAvailableFabricIndex.HasValue())
     {
-        // TODO: Refactor to set mIsPendingFabricDataPresent to true more "directly" when a NOC add/update for
-        //       pending fabric occurs. Can only be done when we have shadow fabric.
-        mIsPendingFabricDataPresent = true;
+        // Mark we have some pending data for a given fabric.
+        fabricIndexToUse = mNextAvailableFabricIndex.Value();
+    }
+    else
+    {
+        // Fabric addition, but adding root would fail on table full: let's not allocate a fabric
+        return CHIP_ERROR_NO_MEMORY;
     }
 
-    return err;
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndexToUse), CHIP_ERROR_INVALID_FABRIC_INDEX);
+    VerifyOrReturnError(SetPendingDataFabricIndex(fabricIndexToUse), CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(mOpCertStore->AddNewTrustedRootCertForFabric(fabricIndexToUse, rcac));
+
+    mStateFlags.Set(StateFlags::kIsPendingFabricDataPresent);
+    mStateFlags.Set(StateFlags::kIsTrustedRootPending);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR FabricTable::FindExistingFabricByNocChaining(FabricIndex pendingFabricIndex, const ByteSpan & noc,
+                                                        FabricIndex & outMatchingFabricIndex) const
+{
+    // Check whether we already have a matching fabric from a cert chain perspective.
+    // To do so we have to extract the FabricID from the NOC and the root public key from the RCAC.
```

Comment:
Can't we extract the root public key from the noc/icac, instead of getting the rcac via `pendingFabricIndex`, and avoid dealing with storage?  But I guess this does work....  It's just very non-obvious why this gives the root that matches the NOC.

Reply:
The root public key is, by definition, only in the root...

Reply:
Ah, I guess the ICAC/RCAC just stores the subject of the signer, which is how you locate the right root?